In [3]:
#lang iracket/lang #:require sicp

# Chapter 1

## 1.1

In [5]:
10
; 10

10

In [6]:
(+ 5 3 4)
; 12

12

In [7]:
(- 9 1)
; 8

8

In [8]:
(/ 6 2)
; 3

3

In [9]:
(+ (* 2 4) (- 4 6))
; 6

6

In [11]:
(define a 3)
;

In [13]:
(define b (+ a 1))
;

In [14]:
(+ a b (* a b))
; 19

19

In [16]:
(= a b)
; #f

#f

In [19]:
(if (and (> b a) (< b (* a b)))
    b
    a)
; 4

4

In [20]:
(cond ((= a 4) 6)
      ((= b 4) (+ 6 7 a))
      (else 25))
; 16

16

In [21]:
(+ 2 (if (> b a) b a))
; 6

6

In [22]:
(* (cond ((> a b) a)
         ((< a b) b)
         (else -1))
   (+ a 1))
; 16

16

## 1.2

In [25]:
(/ (+ 5 
      4 
      (- 2 (- 3 (+ 6 (/ 4 5)))))
   (* 3 
      (- 6 2) 
      (- 2 7)))

-37/150

## 1.3

In [54]:
(define (square x) (* x x))

(define (sum-of-squares a b)
  (+ (square a) (square b)))

(define (larger-of-two a b)
  (if (> a b) 
      a 
      b))

(define (sum-of-squares-of-two-largest a b c)
  (if (> a b) 
      (sum-of-squares a (larger-of-two b c))
      (sum-of-squares b (larger-of-two a c))))

### Tests

In [55]:
(sum-of-squares-of-two-largest 4 2 5)
; 4 * 4 + 5 * 5 = 16 + 25 = 41

41

In [56]:
(sum-of-squares-of-two-largest 3 2 1)
; 3 * 3 + 2 * 2 = 9 + 4 = 13

13

In [57]:
(sum-of-squares-of-two-largest 6 10 7)
; 10 * 10 + 7 * 7 = 100 + 49 = 149

149

## 1.4

In [58]:
(define (a-plus-abs-b a b)
  ((if (> b 0) + -) a b))

The above function performs _a + |b|_

It evaluates a conditional statement to determine whether to use the operator + or - for the function. The if-conditional evaluates `(> b 0)`and returns + for `#t` and - otherwise.

## 1.5

```
(define (p) (p))

(define (test x y)
  (if (= x 0)
      0
      y))

(test 0 (p))
```

In applicative-order evaluation, this test would stall forever as it would be trying to evaluate all the operands first. Function `p` calls itself and never exits.

In normal order evaluation, it would fully expand the function until only primitive operators remain and only then reduce the required operands. This would return 0.

## 1.6

```
(define (sqrt-iter guess x)
  (new-if (good-enough? guess x)
          guess
          (sqrt-iter (improve guess x)
                     x)))
```

If Alyssa attempts to run the above code for calculating a square root, it will get stuck in an infinite recursive loop and never terminate. The reasoning for this is that the built-in `if` is a special form that evaluates differently from normal procedures - it evaluates the predicate and then evaluates the corresponding consequent expression. 

However, the `new-if` procedure obeys the standard rules of applicative-order evaluation by first evaluating all the operands. Since the last operand is another call to `sqrt-iter`, it will start a new procedure call and it will keep looping forever.

## 1.7

In [48]:
(define (square x) (* x x))

(define (good-enough? guess x)
  (< (abs (- (square guess) x)) 0.001))

(define (average x y)
  (/ (+ x y) 2))

(define (improve guess x)
  (average guess (/ x guess)))

(define (sqrt-iter guess x)
  (if (good-enough? guess x)
      guess
      (sqrt-iter (improve guess x)
                 x)))

(define (sqrt x)
  (sqrt-iter 1.0 x))

For very small numbers, the `good-enough?` check provided by the book is inadequate for really small numbers because the allowed difference is too high.

For example, consider taking the square root of 0.0001:

In [57]:
(sqrt 0.0001)

0.03230844833048122

The true square root of 0.0001 is 0.01 however we receive the answer 0.0323. The `good-enough?` procedure takes the square of 0.0323, which is roughly 0.00104, subtracts the radicand from it and checks if the absolute value (0.00104 - 0.001 = 0.00004) is within our error tolerance of 0.001, but our error tolerance is too high to deal with such small numbers.

For large numbers, having the error tolerance be at 0.001 will mean that for sufficiently large numbers, the program will usually never terminate. Due to how floating point numbers work, we run out of precision and the rounding errors make consecutive numbers have a larger gap than 0.001.

"An alternative strategy for implementing good-enough? is to watch how guess changes from one iteration to the next and to stop when the change is a very small fraction of the guess. Design a square-root procedure that uses this kind of end test. Does this work better for small and large numbers?"

In [50]:
(define (new-good-enough? prev-guess guess)
  (< (abs (- guess prev-guess)) 0.00000000001))

(define (new-sqrt-iter guess x)
  (if (new-good-enough? guess (improve guess x))
      guess
      (new-sqrt-iter (improve guess x)
                 x)))

(define (new-sqrt x)
  (new-sqrt-iter 1.0 x))

In [51]:
(new-sqrt 123456789012345)

11111111.061111081

In [30]:
(new-sqrt 0.0001)

0.01

## 1.8

In [52]:
(define (improve guess x)
  (/ (+ (/ x (square guess))
        (* 2 guess))
     3))

(define (cube-root-iter guess x)
  (if (new-good-enough? guess (improve guess x))
      guess
      (cube-root-iter (improve guess x)
                      x)))

(define (cube-root x) (cube-root-iter 1.0 x))

In [39]:
(cube-root 729)
; cube of 9

9.0

In [40]:
(cube-root 125)
; cube of 5

5.0

In [53]:
(cube-root 166.375)
; cube of 5.5

5.500000000000007